App dev proc (#200)
madewithkoji/docs: src/asciidoc/docs/develop/app-dev-process.adoc

For example, you might start from a https://withkoji.com/create/for-developers[developer scaffold] that implements your favorite JavaScript framework, or from an app that already uses some desired functionality, such as a backend database.

. To create a new project, fork an existing app that allows developers to fork it.

.. In the App Store, find your desired starting app and click btn:[View].
+
TIP: To explore the source code for an app before forking it, click menu:...[Advanced Details > Source Code].
If this option is not available, the app does not allow developers to fork it.

.. In the app's detail page, click menu:...[Fork].
+
A copy of the project is created in your Koji account.

. Decide how you want to edit your code.
* *Online* – The Koji code editor is a browser-based integrated development environment (IDE) with built-in development, administration, and testing tools for Koji apps.
+
Learn more: <<editor#>>
* *Locally* – The Koji platform supports working locally in your preferred suite of development tools.
+
Learn more: <<local-development#>>

== [.callout2]#Develop your app#

At their core, Koji apps are full-stack, modern JavaScript applications.

> They are designed to run in webviews, so that they are instantly available on all social platforms and every device.

Author: Definitely a term to add to your list :) I had found a similar resource and used the capitalization you suggest for a previous document. At the time, Sean thought the generic would be more readable, so I changed it ([link to the discussion, for reference](https://github.com/madewithkoji/docs/pull/174#pullrequestreview-667366107)). I think either form would be fine, as long as we standardize. So, if Sean agrees, I'll change it in both places. Otherwise, I'll keep it and you can discuss in future style standards meetings.

> They are designed to run in webviews, so that they are instantly available on all social platforms and every device.

Reviewer: One suggestion might be to avoid `WebView`/`webview` (which feels like it refers more to the implementation inside the app - e.g., `WebView` is a class name on Android) and use something like `in-app browser`/`in-app browsers`. Seems like the phrase is common: https://www.google.com/search?q=in-app+browser

> They are designed to run in webviews, so that they are instantly available on all social platforms and every device.

Author: Updated both pages this way. Thanks!

+
Learn more: <<projects#remixData, remixData key>>, <<projects#@@initialTransform, @@initialTransform key>>

. Configure the entitlements for the Koji platform features in your app.
+
Learn more: <<entitlements#>>

=== Add @withkoji/core to the frontend

Follow these steps to use the @withkoji/core package in the frontend of your app.

[TIP]
====
Before you start developing your code, check for outdated dependencies in the forked app.

* If the forked app uses deprecated Koji packages, migrate your project to use the @withkoji/core package.
+
 Learn more: <<migrate-koji-core#, Do I need to migrate?>>

* Check for other outdated dependencies in the forked app, and update them if necessary.
+
 Learn more: <<work-locally#_installing_the_dependencies, Installing dependencies>>
====

. In your frontend directory, install or update the @withkoji/core package.
+
[source, Bash]
npm install @withkoji/core

. In your frontend code, import the package.
+
[source,Javascript]
import Koji from '@withkoji/core';

. Use the `Koji.config` function to initialize the package with your app's configuration data.
+

> Ideally, your code will initialize the package one time, before any data in the application is rendered.

Reviewer 2: Suggestion: Combine this with the previous sentence; i.e.:

> . Use the `Koji.config` function to initialize the package with your configuration data before any data in the application is rendered.

For example:
+
[source,Javascript]
----
// Initialize
Koji.config(require('././koji.json'));

// render
render();
----

. Use the `Koji.ready` function to indicate that the application is ready to start receiving events.

. Implement the desired core functionality.
For example, you can provide customization options so creators can add their own content and branding to the app before adding it to their profile.
+
Learn more: <<core-frontend-overview#>>, <<capture-user-input#>>, <<thumbnail-image#>>

=== Add @withkoji/core to the backend

If your app has a backend service, follow these steps to use the @withkoji/core package in it.

. In your backend directory, install or update the @withkoji/core package.
+
[source, Bash]
npm install @withkoji/core

> . Make sure the `koji.json` file will be bundled with the backend service during deployment.

Reviewer 2: This begs the question: "How do I verify that it will be bundled?"

+
TIP: In most frameworks, this file won’t be bundled by default because it is not in the same folder as the backend code.
If the file is not bundled, the backend service will not be initialized correctly and the deployment will fail.

> To solve this problem, you can use a tool such as https://www.npmjs.com/package/babel-plugin-inline-json-import[babel-plugin-inline-json-import].

Reviewer 2: Not for adding to this topic, but these are questions that come to mind:
* Why do they need an external tool to do this?
* How do they use this tool to do this?
* Why do we recommend this specific tool?
* What are other alternatives?

I tend to veer away from recommending third-party apps/solutions unless they can be vetted.

Author: As I understand it, we are recommending a file structure that has a potential gotcha for most frameworks, and all apps will need to use something to build their code. This problem is probably less important to call out when someone is using an internally developed app or scaffold as their starting point, because those have already implemented this solution. I'll try to clarify this step a bit, and maybe Sean will have additional feedback to answer some of these questions.

= App development process
:page-slug: app-dev-process
:page-description: Overview of how to develop a Koji app, from forking an existing app, to making your app customizable and enabling platform features, to deploying, testing, and distributing it.
:includespath: ../_includes

The Koji platform is designed for developers like you to build and distribute customizable web apps for content creators.

> On Koji, creators can discover your apps, and then add them to their Link in Bio profiles and share them on social media.

Reviewer 2: > On Koji, creators can discover your apps, and then add them to their Link in Bio profiles and share them on social media.
```suggestion
On Koji, creators can discover your apps, add them to their Link in Bio profiles, and share them on social media.
```

For example, you might start from a https://withkoji.com/create/for-developers[developer scaffold] that implements your favorite JavaScript framework, or from an app that already uses some desired functionality, such as a backend database.

. To create a new project, fork an existing app.

.. In the App Store, find your desired starting app and click btn:[View].
+
TIP: To explore the source code for an app before forking it, click menu:...[Advanced Details > Source Code].
If this option is not available, the app is not forkable.

.. In the app's detail page, click menu:...[Fork].
+
A copy of the project is created in your Koji account.

. Decide how you want to edit your code.
* *Online* – The Koji code editor is a browser-based integrated development environment (IDE) with built-in development, administration, and testing tools for Koji apps.
+
Learn more: <<editor#>>

* *Locally* – The Koji platform supports working locally in your preferred suite of development tools.
+
Learn more: <<local-development#>>

== [.callout2]#Develop your app#

> At their core, Koji apps are full-stack, modern JavaScript applications.

Reviewer 2: > At their core, Koji apps are full-stack, modern JavaScript applications.
```suggestion
At their core, Koji apps are full-stack modern JavaScript applications.
```

> They are designed for in-app browsers, so that they are instantly available on all social platforms and every device.

Reviewer 2: ```suggestion
They are designed for in-app browsers, so that they are instantly available on all social platforms and on every device.
```

The <<withkoji-koji-core#,@withkoji/core package>> enables apps to use core features of the Koji platform, so you can build any type of web application and enable creators to customize it.

=== Define configuration data

In your application code, you must initialize the @withkoji/core package with the configuration data that the Koji platform needs to build, deploy, and run your app.
To define this data, most Koji projects include a `koji.json` file at the root of the project directory.

. Define the instructions for building your code and deploying your app on the Koji platform.
+
Learn more: <<projects#_develop, develop key>>, <<projects#_deploy, deploy key>>

> . If the app has customization options for creators, define the default values for those options and any placeholder values for new, customized versions.

Reviewer 2: ```suggestion
. If the app has customization options for creators, define the default values for those options and any placeholder values for new customized versions.
```

For example:
+
[source,Javascript]
----
// Initialize
Koji.config(require('././koji.json'));

// render
render();
----

. Use the `Koji.ready` function to indicate that the application is ready to start receiving events.

. Implement the desired core functionality.
For example, you can provide customization options so creators can add their own content and branding to the app before adding it to their profiles.
+
Learn more: <<core-frontend-overview#>>, <<capture-user-input#>>, <<thumbnail-image#>>

=== Add @withkoji/core to the backend

If your app has a backend service, follow these steps to use the @withkoji/core package in it.

. In your backend directory, install or update the @withkoji/core package.
+
[source, Bash]
npm install @withkoji/core

. In your backend code, import the package and your app's configuration data.
+
[source,Javascript]
----
import { KojiBackend } from '@withkoji/core';

import kojiConfig from '../../koji.json';
----

+
TIP: To initialize the core package in your backend, the `koji.json` file must be bundled with the backend service during deployment of your app.

> Many frameworks won't bundle this file by default because it is not in the same folder as the backend code.

Reviewer 2: What are these frameworks?

Author: Added "development" but it could probably be more specific in the next version.

> To solve this problem, you can use a tool such as https://www.npmjs.com/package/babel-plugin-inline-json-import[babel-plugin-inline-json-import].

Reviewer 2: Maybe generalize the type of tool required, instead of giving a specific example. If the developer of this tool pulls it out of the market or if it stops working, our readers wouldn't know how to find another solution.

Author: Added "to ensure the configuration data is available to your backend service". Again, something to note for future versions.

. Use `KojiBackend.middleware` to initialize the package with your app's configuration data.
+
The middleware manages the scope for backend functionality, so that your code can support your initial app along with all of the custom versions.
It works by parsing the request headers that the Koji platform uses to pass version-specific data to the backend.
+
You must apply this middleware before handling any routes.

> The middleware will transform `res.locals`, and then modules in KojiBackend will accept `res` as a configuration parameter.

Reviewer 2: ```suggestion
The middleware will transform `res.locals`, and modules in KojiBackend will accept `res` as a configuration parameter.
```

Is KojiBackend a class? If so, should it be between backticks?

Author: It probably should. I added the backticks here. There are probably some more places in other docs where they needed to be added. Something to look for when you update other pages.

. To verify that your latest build was deployed, go to your https://withkoji.com/developer/projects[developer portal] and select the app.
Then, click btn:[View Version History].
+
Learn more: <<app-versions#_viewing_version_history, Viewing version history>>

TIP: If you update your app and deploy it again, the new version of the app replaces the existing version at the production URL.

== [.callout4]#Test your app#

The Koji platform provides a variety of tools for testing and debugging apps.

* If you're working in the Koji code editor, you can preview your app in different modes and on different devices.

* If you're working outside the editor, you can use your browser to verify the basic functionality.
Then, you can point the Koji debugger to your localhost so that you can test all app functionality.

Learn more: <<testing-apps#>>, <<editor#_previewing_your_koji_app, Previewing your Koji app>>, <<work-locally#_testing_your_local_deployment, Testing your local deployment>>

== [.callout5]#Distribute your app#

After you develop, deploy, and test your app, you can share it in any of these ways:

* *Share a direct link with specific people.*

> This option lets your collaborators and customers use your app, without allowing the general public to discover it on the Koji platform.

Reviewer 2: ```suggestion
This option lets your collaborators and customers use your app without allowing the general public to discover it on the Koji platform.
```

It is useful for limited releases or testing.
+
To get the production URL of your app, go to your https://withkoji.com/developer/projects[developer portal] and select the app.
Then, click btn:[View Published Koji], and copy the URL.

* *List it in the https://withkoji.com/apps[Koji App Store].*
This option reaches Koji's global audience of content creators and lets them customize and add your app to their profiles.
You can add marketing assets and documentation to attract more users to your app.

> You can also make money by setting a price for creators to use your app or by setting a transaction fee for in-app purchases.

Reviewer 2: ```suggestion
You can also monetize it by setting a price for creators to use your app or by setting a transaction fee for in-app purchases.
```

== [.callout3]#Deploy your app#

When you're finished developing your app, you can deploy it to the Koji platform.
Deployed apps are accessible at a permanent production URL.

. Deploy your app from your development environment.

* If you're working in the Koji code editor, you can publish your app from there.
+
Learn more: <<editor#_publishing_your_koji_app, Publishing your Koji app>>

* If you're working outside the editor, you must push your changes to your project’s origin repository.
+
Learn more: <<publish-locally-developed#>>

+
After you publish the app, the Koji profiler checks for errors, generates metadata about it, and, if it is error-free, enables the platform entitlements.

. To verify that your latest build was deployed, go to your https://withkoji.com/developer/projects[developer portal] and select the app.
Then, click btn:[View Version History].
+
Learn more: <<app-versions#_viewing_version_history, Viewing version history>>

TIP: If you update your app and deploy it again, the new version of the app replaces the existing version at the production URL.
However, your app listing in the App Store is not automatically updated.

> This feature lets you release test versions of updates without affecting a currently listed version in the App Store.

Reviewer: While this is where we're headed, it is not technically the case today. The most recent published version of an app is the one that is available on the App Store. Versions within the App Store are used only for changes to metadata.

Author: Thanks! Removing the last two sentences for now.
